include/babeltrace/ctf-text/durin.h

```diff
@@ -0,0 +1,169 @@
+ * interface, to access basic debug information. Durin is also the
+ * eldest of the seven Fathers of the Dwarves created by the Vala Aulë
+ * in the lore of Tolkien's Middle-earth.
+ */
+
+/*
+ * This structure corresponds to a single compilation unit (CU) for a
+ * given set of debug information (Dwarf_Debug type).
+ */
+struct durin_cu {
+	Dwarf_Debug *dwarf_info;
+	/* Offset in bytes in the DWARF file to beginning of CU header. */
+	Dwarf_Unsigned offset;
+};
+
+/*
+ * This structure represents a single debug information entry (DIE),
+ * within a compilation unit (CU).
+ */
+struct durin_die {
+	struct durin_cu *cu;
+	Dwarf_Die *dwarf_die;
+	/*
+	 * A depth of 0 represents a root DIE, located in the DWARF
+	 * layout on the same level as its corresponding CU entry. Its
+	 * children DIEs will have a depth of 1, and so forth. All
+	 * "interesting" DIEs for the present use case will be located
+	 * at depth 1, however.
+	 */
+	unsigned int depth;
+};
+
+/**
+ * Instantiate a structure to access compile units (CU) from a given
+ * `dwarf_info`.
+ *
+ * @param dwarf_info	Dwarf_Debug instance
+ * @returns		Pointer to the new durin_cu on success,
+ *			NULL on failure.
+ */
+BT_HIDDEN
+struct durin_cu *durin_cu_create(Dwarf_Debug *dwarf_info);
+
+/**
+ * Destroy the given durin_cu instance.
+ *
+ * @param cu	durin_cu instance
+ */
+BT_HIDDEN
+void durin_cu_destroy(struct durin_cu *cu);
+
+/**
+ * Advance the compile unit `cu` to the next one.
+ *
+ * On success, `cu`'s offset is set to that of the current compile
+ * unit in the executable. On failure, `cu` remains unchanged.
+ *
+ * @param cu	durin_cu instance
+ * @returns	0 on success, -1 on failure
+ */
+BT_HIDDEN
+int durin_cu_next(struct durin_cu *cu);
+
+/**
+ * Instantiate a structure to access debug information entries (DIE)
+ * for the given compile unit `cu`.
+ *
+ * @param cu	durin_cu instance
+ * @returns	Pointer to the new durin_die on success,
+ *		NULL on failure.
+ */
+BT_HIDDEN
+struct durin_die *durin_die_create(struct durin_cu *cu);
+
+/**
+ * Destroy the given durin_die instance.
+ *
+ * @param die	durin_die instance
+ */
+BT_HIDDEN
+void durin_die_destroy(struct durin_die *die);
+
+/**
+ * Advance the debug information entry `die` to the next one.
+ *
+ * @param die	durin_die instance
+ * @returns	0 on success, -1 on failure
+ */
+BT_HIDDEN
+int durin_die_next(struct durin_die *die);
+
+/**
+ * Get a DIE's tag.
+ *
+ * On success, the `tag` out parameter is set to the `die`'s tag's
+ * value. On failure it remains unchanged.
+ *
+ * @param die	durin_die instance
+ * @param tag	Out parameter, the DIE's tag value
+ * @returns	0 on success, -1 on failure.
+ */
+BT_HIDDEN
+int durin_die_get_tag(struct durin_die *die, uint16_t *tag);
+
+/**
+ * Get a DIE's name.
+ *
+ * @param die	durin_die instance
+ * @param name	Out parameter, the DIE's name
+ * @returns	0 on succes, -1 on failure
+ */
+BT_HIDDEN
+int durin_die_get_name(struct durin_die *die, char **name);
```

Comment:
Add comment that the caller does not need to free `name`.
Looking at libdwarf's source code seems to indicate that the string is valid as long as the underlying `dwarf_die` is not `free`'d. This also means that `durin_die_next()` must not be called.